### PR TITLE
Add cafeteria occupancy base stats API

### PR DESCRIPTION
**Pull request by avan1235**

### CODE REVIEW

File: django-server/server/api/models.py [hunk 1]
```diff
@@ -124,3 +124,10 @@ def save(self, *args, **kwargs):
     def calculate_new_avg_review(self, old_review_stars, number_of_reviews):
         current_sum = old_review_stars * number_of_reviews + self.stars
         return current_sum / (number_of_reviews + 1)
+
+
+class OccupancyStatsData(object):
+
+    def __init__(self, **kwargs):
+        for field in ('id', 'timestamp_name', 'occupancy', 'occupancy_relative'):
```
Comment (kantoniak) [inline django-server/server/api/models.py:132]: Was it possible to use custom object for occupancy? This is low priority here.

File: django-server/server/api/urls.py [hunk 2]
```diff
@@ -26,6 +26,11 @@
 urlpatterns = [
     path('', include(router.urls)),
     path('', include(cafeterias_router.urls)),
+    path('cafeterias/<int:cafeteria_pk>/stats/hour', HourStatsView.as_view()),
+    path('cafeterias/<int:cafeteria_pk>/stats/day', DayStatsView.as_view()),
+    path('cafeterias/<int:cafeteria_pk>/stats/week', WeekStatsView.as_view()),
+    path('cafeterias/<int:cafeteria_pk>/stats/month', MonthStatsView.as_view()),
+    path('cafeterias/<int:cafeteria_pk>/stats/year', YearStatsView.as_view()),
```
Comment (kantoniak) [inline django-server/server/api/urls.py]: `cafeterias/<int:cafeteria_pk>/occupancy/` with grouping through filters. Once we implement other filters there will be 5 copies of everything.

File: django-server/server/api/views.py [hunk 3]
```diff
@@ -112,6 +112,131 @@ def get_queryset(self):
         return FixedMenuOptionReview.objects.all().filter(option_id=self.kwargs['option_pk']).order_by('id')
 
 
+class TimeStampedFilterSet(filters.FilterSet):
+    timestamp_start = filters.DateTimeFilter(required=True)
```
Comment (kantoniak) [inline django-server/server/api/views.py:116]: Can we add `from` and `to` filters? It seems this comes at no extra cost.

File: django-server/server/api/views.py [hunk 4]
```diff
@@ -112,6 +112,131 @@ def get_queryset(self):
...
+        if getattr(self, 'swagger_fake_view', False):
+            # queryset just for schema generation metadata
+            return []
+        id = self.kwargs.get('cafeteria_pk')
+        count = self.request.query_params.get('count')
+        timestamp_start = self.request.query_params.get('timestamp_start')
+        if id is None or count is None or timestamp_start is None:
+            raise ValueError('required params not specified')
+
+        timestamp_start = parse(timestamp_start)
```
Comment (kantoniak) [inline django-server/server/api/views.py:135]: Please either use two variables or move it to line 132, right after original definition.

File: django-server/server/api/views.py [hunk 5]
```diff
@@ -112,6 +112,131 @@ def get_queryset(self):
...
+        events = CameraEvent.objects.filter(cafeteria_id=id)
+        before = events.filter(timestamp__lte=timestamp_start).order_by('timestamp')
+        overrides = before.filter(event_type=CameraEvent.EventType.OCCUPANCY_OVERRIDE.value)
+        init_override_value = 0 if len(overrides) == 0 else overrides.latest('timestamp').event_value
+        changed = self.count_people(before) if len(overrides) == 0 \
+            else self.count_people(before.filter(timestamp__gt=overrides.latest('timestamp').timestamp))
```
Comment (kantoniak) [inline django-server/server/api/views.py:141]: These lines are quite complex. Can we do something like below?
```
initial_people_count = 0;
events_before = before;

if not overrides:
    initial_people_count = ...
    events_before = ...

changed = self.count_people(events_before);
```

File: django-server/server/api/views.py [hunk 6]
```diff
@@ -112,6 +112,131 @@ def get_queryset(self):
...
+
+
+class YearStatsView(StatsView):
+
+    def get_timestamp_delta(self, datetime: datetime):
+        return timedelta(days=(datetime.replace(year=datetime.year + 1) - datetime).days)
+
+    def get_timestamp_name(self, datetime: datetime):
+        return '{}'.format(datetime.year)
+
```
Comment (kantoniak) [inline django-server/server/api/views.py:238]: See earlier comment about copying.

Comment (avan1235) [reply django-server/server/api/views.py:238]: this is done only once to specify the ranges of filter type so I leave it there

File: django-server/server/api/urls.py (hunk 2 above)
Comment (avan1235) [reply django-server/server/api/urls.py]: Written in more abstract way now, only the interface has to be implemented and neme specified in single place to get stats for all types (they don't have to specified explicitly for every stats type)

File: django-server/server/api/views.py (hunk 4 above)
Comment (avan1235) [reply django-server/server/api/views.py]: fixed

File: django-server/server/api/views.py (hunk 3 above)
Comment (avan1235) [reply django-server/server/api/views.py]: added. the minimum of time_end, time.now(), count is taken into account when stats are generated

File: django-server/server/api/views.py (hunk 5 above)
Comment (avan1235) [reply django-server/server/api/views.py]: rewritten in more abstract way but I think this change is included

File: django-server/server/api/models.py (hunk 1 above)
Comment (avan1235) [reply django-server/server/api/models.py:132]: I don't understand the question here

File: django-server/server/api/views.py [hunk 7]
```diff
@@ -112,6 +114,203 @@ def get_queryset(self):
...
+
+class MonthStatsDivider(StatsDivider):
+    def get_timestamp_delta(self, time: datetime):
+        return timedelta(days=monthrange(time.year, time.month)[1])
+
+
+class YearStatsDivider(StatsDivider):
+    def get_timestamp_delta(self, time: datetime):
+        return timedelta(days=(time.replace(year=time.year + 1) - time).days)
```
[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/api/views.py:148]: it has to be a function as it depends on the current date (for example month/year) how many day s shoft

File: django-server/server/admin/static/admin/stats.js [hunk 8]
```diff
@@ -278,10 +278,19 @@ const App = new Vue({
...
         .then(data => {
+          let mapped_data = data.map(value => {
+              return  {
+                timestamp: value.timestamp,
+                value: {
+                  relative: value.occupancy_relative,
+                  total: value.occupancy
+                }
+              };
+            });
```
[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/admin/static/admin/stats.js:289]: will fix the firs part as I don't impleemnt reviews and don't know the format

File: django-server/server/admin/static/admin/stats.js [hunk 9]
```diff
@@ -662,6 +671,15 @@ function makeOccupancyData(startDate, endDate) {
 }
 
 function loadOccupancy(cafeteria_id, from, to) {
+  return fetch(API_URL_BASE + 'cafeterias/' + cafeteria_id + '/stats/occupancy/by_hour?timestamp_start=' + from.toISOString() + '&timestamp_end=' + to.toISOString())
```
[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/admin/static/admin/stats.js:674]: done

[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/admin/static/admin/stats.js:674]: done

File: django-server/server/api/views.py [hunk 10]
```diff
@@ -112,6 +114,203 @@ def get_queryset(self):
...
+        data = self.get_full_queryset(cafeteria_pk)
+        before_data = data.filter(**{lookup_lte: timestamp_start}).order_by(self.timestamp_field_name())
+        count_value = self.init_value(before_data)
+        begin_stamp = timestamp_start
+        end_stamp = timestamp_start + divider.get_timestamp_delta(divider, timestamp_start)
+
+        results = [(count_value, begin_stamp)]
+        for interval_i in range(int(count) - 1):
+            finish_now = False
+            if end_stamp > utc.localize(datetime.now()) or end_stamp > timestamp_end:
```
[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/api/views.py:224]: ok, thanks

File: django-server/server/admin/static/admin/stats.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (avan1235) [reply django-server/server/admin/static/admin/stats.js:289]: the mapping has to be where it si now as loadOccupancy return promise, not a value. Visually it still not works

File: django-server/server/api/views.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kantoniak) [reply django-server/server/api/views.py:148]: Then map can be label to function (not method) and code can call the function. We still don't need a class for each of them.

File: django-server/server/api/views.py [hunk 11]
```diff
@@ -197,13 +192,14 @@ def get_queryset(self):
...
-        end_string = self.request.query_params.get('timestamp_end')
-        divider = self.divider
+        start_string = self.request.query_params.get('start_timestamp')
+        end_string = self.request.query_params.get('end_timestamp')
+        group_by_string = self.request.query_params.get('group_by')
+        divider = get_divider(group_by_string)()
 
-        timestamp_start = datetime.min + divider.get_timestamp_delta(divider, datetime.min) \
+        timestamp_start = datetime.min + divider.get_timestamp_delta(datetime.min) \
             if start_string is None else timestamp_parse(start_string)
```
Comment (kantoniak) [inline django-server/server/api/views.py:201]: These inline conditional statements add complexity to code. Try not to use them in future.

Comment (kantoniak) [reply django-server/server/api/views.py:201]: You can run `flake8-cognitive-complexity` locally to test code for such structures.

File: django-server/server/api/views.py [hunk 12]
```diff
@@ -301,7 +290,7 @@ def next_count_value(self, count_value, curr_queryset: QuerySet):
     def map_to_result_objects(self, index, value, stamp, cafeteria_pk):
```
Comment (kantoniak) [inline django-server/server/api/views.py:290]: s/stamp/timestamp/

File: django-server/server/api/views.py (hunk 11 above)
Comment (avan1235) [reply django-server/server/api/views.py:201]: wil fix

File: django-server/server/api/views.py (hunk 12 above)
Comment (avan1235) [reply django-server/server/api/views.py:296]: fixed

**Pull request merged**